airbrakes/data_handling/data_processor.py

```diff
@@ -202,52 +202,111 @@ def _calculate_current_altitudes(self) -> npt.NDArray[np.float64]:
+                x_accel = compensated_accelerations[idx][0]
+                y_accel = compensated_accelerations[idx][1]
+                z_accel = compensated_accelerations[idx][2]
+                # Angular rates are in rads/s
+                gyro_x = data_packet.estAngularRateX
+                gyro_y = data_packet.estAngularRateY
+                gyro_z = data_packet.estAngularRateZ
+
+                # Check for missing data points
+                if any(val is None for val in [gyro_x, gyro_y, gyro_z]):
+                    return rotated_accelerations
+
+                # scipy docs for more info: https://docs.scipy.org/doc/scipy/reference/generated/scipy.spatial.transform.Rotation.html
+                # Calculate the delta quaternion from the angular rates
+                delta_rotation = R.from_rotvec([gyro_x * dt, gyro_y * dt, gyro_z * dt])
+
+                # Update the current orientation by applying the delta rotation
+                current_orientation = current_orientation * delta_rotation
+
+                # Rotate the acceleration vector using the updated orientation
+                rotated_accel = current_orientation.apply([x_accel, y_accel, z_accel])
+                # Vertical acceleration will always be the 3rd element of the rotated vector,
+                # regardless of orientation. For simplicity, we multiply by -1 so that acceleration
+                # during motor burn is positive, and acceleration due to drag force during coast
+                # phase is negative.
+                rotated_accelerations[idx] = -rotated_accel[2]
+
+            # Update the class attribute with the latest quaternion orientation
+            self._current_orientation_quaternions = current_orientation
+
+            return rotated_accelerations
+
+        # the quaternions are not uncertain, so we can directly use the quaternions given from
+        # the IMU.
+        current_orientation = np.array(
+            [
+                last_packet.estOrientQuaternionW,
+                last_packet.estOrientQuaternionX,
+                last_packet.estOrientQuaternionY,
+                last_packet.estOrientQuaternionZ,
+            ]
+        )
+        # rotate each compensated acceleration by the most recent quaternion orientation. This
+        # introduces some slight error, because the orientation for the last packet is not going
+        # to be the orientation for each packet, but it should be a negligible difference.
+        current_orientation = R.from_quat(current_orientation, scalar_first=True)
+        rotated_accels = current_orientation.apply(compensated_accelerations)
 
         # Update the class attribute with the latest quaternion orientation
         self._current_orientation_quaternions = current_orientation
-
-        return rotated_accelerations
+        # we only want to return the negative vertical component of each acceleration
+        return np.array([-accel[2] for accel in rotated_accels])
```

Comment:
```suggestion
        return -rotated_accels[:, -1]
```
This is **95.1%** faster since it's vectorized!!

```python
import timeit

setup = """
import numpy as np

n = np.array([[1.213, 1.213 + i, 1.213 +2*i] for i in range(100)])
"""

test_1 = """
b = -n[:, -1]
"""

test_2 = """
b = np.array([-j[-1] for j in n])
"""

time_1 = timeit.timeit(test_1, setup, number=1000000)
time_2 = timeit.timeit(test_2, setup, number=1000000)

print(time_1)
print(time_2)
```
```bash
└─[$] <git:(adjust-queue-fetching*)> python scripts/perf.py
0.6513239109990536
24.606471167000564
```

Reply:
man I was sure there was a way to vectorize this but couldn't figure it out lol

Reply:
https://numpy.org/devdocs/user/basics.rec.html#structured-arrays

oh my god look what I found

Reply:
if we did this, would msgspec not be needed?

Reply:
it's a bit early to say that, but I'm looking at alternative ways of representing data, like pandas Dataframes (which are also labeled). It's just that it hurts the readability of having data packets to represent data v/s Dataframes, but the latter is way more suitable for us since we keep working on the columns (e.g. data processor and apogee predictor) rather than the row (only used in logging)

Reply:
Resolved in #112